docs/Software Engineering Blog/posts/spring-boot-3-1-docker-compose-support-intuitive-consistent-develop-run-tests-containers.md

```diff
@@ -0,0 +1,229 @@
+Both need extra effort to execute and remember (engineers are extremely lazy).
+Don't even mention the integration difficulty in unit tests.
+
+### In Testing
+
+When a test needs to cover the dependency, it first comes up with the mock framework in a programmatic way.
+But it turns out to lose so many details because of the lack of real integration. such as
+
+1. The dialect and SQL execution difference between in-memory h2 DB and a real MySQL DB.
+2. Provide a mock Java object to make an API Client test lose the coverage of the JSON parsing process.
+
+so some of them start to use the TestContainer technic, to provide a real instance during unit tests. However,
+TestContainer suffers from some points too, like:
+
+1. need [support library](https://java.testcontainers.org/modules/databases/mysql/) to gain more interaction.
+2. a programmatic way to set up, which is different from the development time (docker-compose way).
+3. extra setting to share containers in different tests
+
+## Introducing the Spring-Boot-Docker-Compose
+
+In conclusion, the way of providing dependency containers isn't a flawless practice for now. So Spring community
+provides a new way to seamlessly integrate with Spring Boot App and docker-compose
+called [Spring-Boot-Docker-Compose](https://docs.spring.io/spring-boot/reference/features/dev-services.html#features.dev-services.docker-compose),
+follow the starting guide could easily run the app with the dependency container together in one command.
+
+just add the dependency
+
+```groovy
+developmentOnly 'org.springframework.boot:spring-boot-docker-compose'
+```
+
+then provide a `compose.yaml` file in the root folder can do the trick. If the compose.yaml has a different name or in a
+different folder, `spring.docker.compose.file` can solve the problem, refer
+to [Using a Specific Compose File](https://docs.spring.io/spring-boot/reference/features/dev-services.html#features.dev-services.docker-compose.specific-file).
+
+![img.png](resources/sb-docker/start_with_docker_console.png)
+
+![start_with_docker_podman.png](resources/sb-docker/start_with_docker_podman.png)
+
+From now on, every project member can run the app without manually starting the dependency component container
+cluster.
+
+## Little More Setup to Make It Better
+
+To enhance the development and testing experience, consider the following improvements.
+
+### For Development
+
+#### Docker Profile Connected to Spring Profile:
+
+Since the practice of spring profiles is usually used as a dependency management method. For example, we add a
+profile `redis` in `spring.profiles.include`, to include `application-redis.yaml`, which contains the related config for
+using redis.
+
+So we want to link docker compose profiles to Spring profiles. In this way, developers can ensure that the correct
+configuration is used and the corresponding container is up also.
+
+To do that, we need some modification:
```

Comment:
⚠️ **[vale]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>[Google.We] Try to avoid using first-person plural like 'we'.